aiida_quantumespresso/calculations/__init__.py

```diff
@@ -82,6 +122,29 @@ def define(cls, spec):
         spec.input_namespace('pseudos', valid_type=orm.UpfData, dynamic=True,
             help='A mapping of `UpfData` nodes onto the kind name to which they should apply.')
 
+        _allowed_parallelization_mapping = {
+            flag_name: (default, help_) for flag_name, default, help_ in cls._ALLOWED_PARALLELIZATION_FLAGS
+        }
+        for flag_name in cls._ENABLED_PARALLELIZATION_FLAGS:
+            try:
+                default, help_ = _allowed_parallelization_mapping[flag_name]
+            except KeyError as exc:
+                raise KeyError(
+                    "The parallelization flag '{}' specified in _ENABLED_PARALLELIZATION_FLAGS ".format(flag_name) +
+                    'does not exist in _ALLOWED_PARALLELIZATION_FLAGS. Please report this issue '
+                    'to the calculation plugin developer.'
+                ) from exc
+
+            if default is None:
+                extra_kwargs = {'required': False}
+            else:
+                # We use 'functools.partial' here because lambda: `orm.Int(default)` would
+                # bind to the name `default`, and look up its value only when the lambda is
+                # actually called. Because `default` changes throughout the for-loop, this
+                # would give the wrong value.
+                extra_kwargs = {'default': partial(orm.Int, default)}
```

Comment:
Not sure about specifying these defaults. I think this would mean that all these calculations will now always have a bunch of int nodes as inputs even if the user doesn't specify anything. I am not sure we want this, at least not until we have proper automatic node deduplication in the database. If we still need the defaults to be set for consistency, we can maybe just do this on `prepare_for_submission` of the base class, but without going through input nodes

Reply:
I added these defaults mostly for caching: Two calculations could otherwise have different inputs but be functionally equivalent. They wouldn't use caching, just because the inputs are different.

I'll leave it up to you to decide if that's an important reason - I can also see the argument that the extra input nodes are unnecessary. Otherwise I can just remove the defaults.

Reply:
Hmm, didn't think about the caching. Then again, maybe if we miss a cache hit because of the parallelization flags is not too bad as it should not occur too often. That reminds me though: didn't you mention somewhere that the actual defaults in QE are a bit surprising, i.e. not all 1 or something like that?

Reply:
The QE defaults are 1 for all inputs _except_ `ndiag` - that's why we don't have a default for that here. QE determines it based on the number of available processors and some of the other parallelization levels.

Should I remove the defaults for the other parallelization options, then?

Reply:
Right, I'm asking whether the _code_ allowing for defaults should also be removed.

Reply:
> the default is not None code path is dead now.

Not quite. A subclass could still override `_PARALLELIZATION_FLAGS` to define default for one or more flags and automatically get the spec to work correctly. Not sure if this would be useful, but could be the case. So maybe keeping the code is ok. That being said, what is this `MappingProxyType` and why do you use that to create the class attribute?